### ISSUE TITLE
[x64\windows] Assertion failed '!foundDiff' in 'System.Runtime.Intrinsics.X86.Pclmulqdq:CarrylessMultiply

### ISSUE THREAD

**Issue by AriNuer** [dotnet/coreclr#23089]
**Job:**
https://ci.dot.net/job/dotnet_coreclr/job/master/view/x64/job/jitstress/job/x64_checked_windows_nt_jitx86hwintrinsicnosimd/384/#showFailuresLink

**Failed tests:**
JIT_HardwareIntrinsics._X86_Sse2_Sse2_r_Sse2_r_._X86_Sse2_Sse2_r_Sse2_r_cmd
JIT_HardwareIntrinsics._X86_Fma_Vector128_Fma_r_Fma_r_._X86_Fma_Vector128_Fma_r_Fma_r_cmd
JIT_HardwareIntrinsics._X86_Sse3_Sse3_ro_Sse3_ro_._X86_Sse3_Sse3_ro_Sse3_ro_cmd
JIT_HardwareIntrinsics._X86_Sse41_MinHorizontal_r_MinHorizontal_r_._X86_Sse41_MinHorizontal_r_MinHorizontal_r_cmd
JIT_HardwareIntrinsics._X86_Sse_Sse_ro_Sse_ro_._X86_Sse_Sse_ro_Sse_ro_cmd
JIT_HardwareIntrinsics._X86_Avx_ConvertToVector_r_ConvertToVector_r_._X86_Avx_ConvertToVector_r_ConvertToVector_r_cmd
JIT_HardwareIntrinsics._X86_Avx_Vector128_Avx_ro_Avx_ro_._X86_Avx_Vector128_Avx_ro_Avx_ro_cmd
JIT_HardwareIntrinsics._X86_Sse41_Sse41_r_Sse41_r_._X86_Sse41_Sse41_r_Sse41_r_cmd
JIT_HardwareIntrinsics._X86_Avx_ConvertToVector_ro_ConvertToVector_ro_._X86_Avx_ConvertToVector_ro_ConvertToVector_ro_cmd
JIT_HardwareIntrinsics._X86_Sse41_ConvertToVector128_r_ConvertToVector128_r_._X86_Sse41_ConvertToVector128_r_ConvertToVector128_r_cmd
JIT_HardwareIntrinsics._X86_Avx2_Vector128_Avx2_ro_Avx2_ro_._X86_Avx2_Vector128_Avx2_ro_Avx2_ro_cmd
JIT_HardwareIntrinsics._X86_Sse_X64_Sse_X64_r_Sse_X64_r_._X86_Sse_X64_Sse_X64_r_Sse_X64_r_cmd
JIT_HardwareIntrinsics._X86_Sse_X64_Sse_X64_ro_Sse_X64_ro_._X86_Sse_X64_Sse_X64_ro_Sse_X64_ro_cmd
JIT_HardwareIntrinsics._X86_Sse41_MultipleSumAbsoluteDifferences_r_MultipleSumAbsoluteDifferences_r_._X86_Sse41_MultipleSumAbsoluteDifferences_r_MultipleSumAbsoluteDifferences_r_cmd
JIT_HardwareIntrinsics._X86_Aes_Aes_ro_Aes_ro_._X86_Aes_Aes_ro_Aes_ro_cmd
JIT_HardwareIntrinsics._X86_Avx_Avx_ro_Avx_ro_._X86_Avx_Avx_ro_Avx_ro_cmd
JIT_HardwareIntrinsics._X86_Sse41_X64_Sse41_X64_ro_Sse41_X64_ro_._X86_Sse41_X64_Sse41_X64_ro_Sse41_X64_ro_cmd
JIT_HardwareIntrinsics._X86_Avx_Vector128_Avx_r_Avx_r_._X86_Avx_Vector128_Avx_r_Avx_r_cmd
JIT_HardwareIntrinsics._X86_Sse2_X64_Sse2_X64_ro_Sse2_X64_ro_._X86_Sse2_X64_Sse2_X64_ro_Sse2_X64_ro_cmd
JIT_HardwareIntrinsics._X86_Sse42_Sse42_ro_Sse42_ro_._X86_Sse42_Sse42_ro_Sse42_ro_cmd
JIT_HardwareIntrinsics._X86_Ssse3_Ssse3_r_Ssse3_r_._X86_Ssse3_Ssse3_r_Ssse3_r_cmd
JIT_HardwareIntrinsics._X86_Sse3_Sse3_r_Sse3_r_._X86_Sse3_Sse3_r_Sse3_r_cmd
JIT_HardwareIntrinsics._X86_Sse41_MinHorizontal_ro_MinHorizontal_ro_._X86_Sse41_MinHorizontal_ro_MinHorizontal_ro_cmd
JIT_HardwareIntrinsics._X86_Avx2_Avx2_r_Avx2_r_._X86_Avx2_Avx2_r_Avx2_r_cmd
JIT_HardwareIntrinsics._X86_Aes_Aes_r_Aes_r_._X86_Aes_Aes_r_Aes_r_cmd
JIT_HardwareIntrinsics._X86_Sse_Sse_r_Sse_r_._X86_Sse_Sse_r_Sse_r_cmd
JIT_HardwareIntrinsics._X86_Sse2_X64_Sse2_X64_r_Sse2_X64_r_._X86_Sse2_X64_Sse2_X64_r_Sse2_X64_r_cmd
JIT_HardwareIntrinsics._X86_Fma_Vector128_Fma_ro_Fma_ro_._X86_Fma_Vector128_Fma_ro_Fma_ro_cmd
JIT_HardwareIntrinsics._X86_Sse41_Sse41_ro_Sse41_ro_._X86_Sse41_Sse41_ro_Sse41_ro_cmd
JIT_HardwareIntrinsics._X86_Sse41_MultipleSumAbsoluteDifferences_ro_MultipleSumAbsoluteDifferences_ro_._X86_Sse41_MultipleSumAbsoluteDifferences_ro_MultipleSumAbsoluteDifferences_ro_cmd
JIT_HardwareIntrinsics._X86_Avx2_Avx2_ro_Avx2_ro_._X86_Avx2_Avx2_ro_Avx2_ro_cmd
JIT_HardwareIntrinsics._X86_Pclmulqdq_Pclmulqdq_ro_Pclmulqdq_ro_._X86_Pclmulqdq_Pclmulqdq_ro_Pclmulqdq_ro_cmd
JIT_HardwareIntrinsics._X86_Avx2_Vector128_Avx2_r_Avx2_r_._X86_Avx2_Vector128_Avx2_r_Avx2_r_cmd
JIT_HardwareIntrinsics._X86_Sse42_Sse42_r_Sse42_r_._X86_Sse42_Sse42_r_Sse42_r_cmd
JIT_HardwareIntrinsics._X86_Sse41_X64_Sse41_X64_r_Sse41_X64_r_._X86_Sse41_X64_Sse41_X64_r_Sse41_X64_r_cmd
JIT_HardwareIntrinsics._X86_Sse41_ConvertToVector128_ro_ConvertToVector128_ro_._X86_Sse41_ConvertToVector128_ro_ConvertToVector128_ro_cmd
JIT_HardwareIntrinsics._X86_Avx_Avx_r_Avx_r_._X86_Avx_Avx_r_Avx_r_cmd
JIT_HardwareIntrinsics._X86_Pclmulqdq_Pclmulqdq_r_Pclmulqdq_r_._X86_Pclmulqdq_Pclmulqdq_r_Pclmulqdq_r_cmd
JIT_HardwareIntrinsics._X86_Sse2_Sse2_ro_Sse2_ro_._X86_Sse2_Sse2_ro_Sse2_ro_cmd
JIT_HardwareIntrinsics._X86_Ssse3_Ssse3_ro_Ssse3_ro_._X86_Ssse3_Ssse3_ro_Ssse3_ro_cmd

**Failed log:**
```
JIT_HardwareIntrinsics._X86_Sse2_Sse2_r_Sse2_r_._X86_Sse2_Sse2_r_Sse2_r_cmd [FAIL]
        
        Assert failure(PID 8972 [0x0000230c], Thread: 7040 [0x1b80]): Assertion failed '!foundDiff' in 'System.Runtime.Intrinsics.X86.Sse2:Add(struct,struct):struct' (IL size 8)
        
            File: d:\j\workspace\x64_checked_w---2304cfbe\src\jit\lsra.cpp Line: 2112
            Image: D:\j\workspace\x64_checked_w---2304cfbe\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root\CoreRun.exe
        
        
  
  Return code:      1
  Raw output file:      D:\j\workspace\x64_checked_w---2304cfbe\bin\tests\Windows_NT.x64.Checked\Reports\JIT.HardwareIntrinsics\X86\Sse2\Sse2_r\Sse2_r.output.txt
  Raw output:
  BEGIN EXECUTION
         "D:\j\workspace\x64_checked_w---2304cfbe\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root\corerun.exe" Sse2_r.exe 
        Supported ISAs:
          AES:       False
          AVX:       False
          AVX2:      False
          BMI1:      True
          BMI2:      True
          FMA:       False
          LZCNT:     True
          PCLMULQDQ: False
          POPCNT:    True
          SSE:       False
          SSE2:      False
          SSE3:      False
          SSE4.1:    False
          SSE4.2:    False
          SSSE3:     False
        
        Beginning test case Add.Double at 3/6/2019 5:26:36 AM
        Random seed: 20010415; set environment variable CORECLR_SEED to this value to repro
        
        Beginning scenario: RunUnsupportedScenario
        Beginning scenario: RunBasicScenario_UnsafeRead
        Expected: 100
        Actual: -1073740286
        END EXECUTION - FAILED
        FAILED
        Test Harness Exitcode is : 1
        
  To run the test:
  > set CORE_ROOT=D:\j\workspace\x64_checked_w---2304cfbe\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root
  > D:\j\workspace\x64_checked_w---2304cfbe\bin\tests\Windows_NT.x64.Checked\JIT\HardwareIntrinsics\X86\Sse2\Sse2_r\Sse2_r.cmd
  
        Expected: True
        Actual:   False
        Stack Trace:
          D:\j\workspace\x64_checked_w---2304cfbe\bin\tests\Windows_NT.x64.Checked\TestWrappers\JIT.HardwareIntrinsics\JIT.HardwareIntrinsics.XUnitWrapper.cs(12079,0): at JIT_HardwareIntrinsics._X86_Sse2_Sse2_r_Sse2_r_._X86_Sse2_Sse2_r_Sse2_r_cmd()
```

**Comment by tannergooding** [comment]
@CarolEidt, looks like this failure is also caused by #22791.

It is a different assert from https://github.com/dotnet/coreclr/issues/23059 and repros in x64 rather than x86.

It also requires both TieredCompilation=0 and FeatureSIMD=0 to reproduce.

**Comment by tannergooding** [comment]
The only thing that pops out as possibly related is the `asgType = impNormStructType(structHnd);` call, which will do different things between `FeatureSIMD=1` and `FeatureSIMD=0`

**Comment by CarolEidt** [comment]
> The only thing that pops out as possibly related is the asgType = impNormStructType(structHnd); call, which will do different things between FeatureSIMD=1 and FeatureSIMD=0

Yes, I'm nearly certain that's the source of both issues, and if I can make a little progress on a couple of other fronts I'm hoping it isn't too hard to fix (I believe I was mistaken in thinking, for my fix in #22791, that it is OK to have a comma on the rhs of a struct assignment.)

**Comment by CarolEidt** [comment]
#23141 fixed this as well as #23059
